drivers/i2s/i2s_nrf_tdm.c

```diff
@@ -0,0 +1,1216 @@
+			int ret = k_msgq_put(&drv_data->rx_queue, &buf, K_NO_WAIT);
+
+			if (ret < 0) {
+				LOG_ERR("No room in RX queue");
+				drv_data->state = I2S_STATE_ERROR;
+				stop_transfer = true;
+
+				free_rx_buffer(drv_data, &buf);
+			} else {
+
+				/* If the TX direction is not active and
+				 * the transfer should be stopped after
+				 * the current block, stop the reception.
+				 */
+				if (drv_data->active_dir == I2S_DIR_RX && drv_data->stop) {
+					drv_data->discard_rx = true;
+					stop_transfer = true;
+				}
+			}
+		}
+	}
+
+	if (released->p_tx_buffer) {
+		buf.mem_block = (void *)released->p_tx_mem_slab;
+		buf.dmm_buf = (void *)released->p_tx_buffer;
+		/* If the last buffer that was to be transferred has just been
+		 * released, it is time to stop the transfer.
+		 */
+		if (released->p_tx_buffer == drv_data->last_tx_buffer) {
+			drv_data->discard_rx = true;
+			stop_transfer = true;
+		} else {
+			free_tx_buffer(drv_data, &buf);
+		}
+	}
+
+	if (stop_transfer) {
+		tdm_stop(drv_cfg->p_reg);
+	} else if (status & NRFX_TDM_STATUS_NEXT_BUFFERS_NEEDED) {
+		tdm_buffers_t next = {0};
+
+		if (drv_data->active_dir != I2S_DIR_RX) { /* -> TX active */
+			if (drv_data->stop) {
+				/* If the stream is to be stopped, don't get
+				 * the next TX buffer from the queue, instead
+				 * supply the one used last time (it won't be
+				 * transferred, the stream will stop right
+				 * before this buffer would be started again).
+				 */
+				next.p_tx_buffer = drv_data->last_tx_buffer;
+				next.p_tx_mem_slab = drv_data->last_tx_mem_slab;
+				next.buffer_size = 1;
+			} else if (get_next_tx_buffer(drv_data, &next)) {
+				/* Next TX buffer successfully retrieved from
+				 * the queue, nothing more to do here.
+				 */
+			} else if (drv_data->state == I2S_STATE_STOPPING) {
+				/* If there are no more TX blocks queued and
+				 * the current state is STOPPING (so the DRAIN
+				 * command was triggered) it is time to finish
+				 * the transfer.
+				 */
+				drv_data->stop = true;
+				/* Supply the same buffer as last time; it will
+				 * not be transferred anyway, as the transfer
+				 * will be stopped earlier.
+				 */
+				next.p_tx_buffer = drv_data->last_tx_buffer;
+				next.p_tx_mem_slab = drv_data->last_tx_mem_slab;
+				next.buffer_size = 1;
+			} else {
+				/* Next TX buffer cannot be supplied now.
+				 * Defer it to when the user writes more data.
+				 */
+				drv_data->next_tx_buffer_needed = true;
+				return;
+			}
+		}
+		(void)supply_next_buffers(drv_data, &next);
+	}
+}
+
+static void clock_manager_init(const struct device *dev)
+{
+#if CONFIG_CLOCK_CONTROL_NRF
+	clock_control_subsys_t subsys;
+	struct tdm_drv_data *drv_data = dev->data;
+
+#if NRF_CLOCK_HAS_HFCLKAUDIO
+	const struct i2s_nrfx_drv_cfg *drv_cfg = dev->config;
+
+	if (drv_cfg->clk_src == ACLK) {
+		subsys = CLOCK_CONTROL_NRF_SUBSYS_HFAUDIO;
+	} else {
+		subsys = CLOCK_CONTROL_NRF_SUBSYS_HF;
+	}
+#else
+	subsys = CLOCK_CONTROL_NRF_SUBSYS_HF;
```

Comment:
What's the point in getting the clock manager for this subsystem when clock is to be requested only for the ACLK source? See: https://github.com/zephyrproject-rtos/zephyr/blob/f9ccc61906397d350ff63d407180c0201f81f7a3/drivers/i2s/i2s_nrf_tdm.c#L578